packages/jaeger-ui/src/components/TracePage/TraceTimelineViewer/VirtualizedTraceView.tsx

```diff
@@ -358,24 +358,39 @@ export class VirtualizedTraceViewImpl extends React.Component<VirtualizedTraceVi
     const isDetailExpanded = detailStates.has(spanID);
     const isMatchingFilter = findMatchesIDs ? findMatchesIDs.has(spanID) : false;
     const showErrorIcon = isErrorSpan(span) || (isCollapsed && spanContainsErredSpan(trace.spans, spanIndex));
-    const criticalPathSections = criticalPath?.filter(each => {
-      if (isCollapsed) {
-        const allChildSpanIds = [spanID, ...childSpanIds];
-        // This function called recursively to find all descendants of a span
-        const findAllDescendants = (currentChildSpanIds: string[]) => {
-          currentChildSpanIds.forEach(eachId => {
-            const currentChildSpan = trace.spans.find(a => a.spanID === eachId)!;
-            if (currentChildSpan.hasChildren) {
-              allChildSpanIds.push(...currentChildSpan.childSpanIds);
-              findAllDescendants(currentChildSpan.childSpanIds);
-            }
-          });
-        };
-        findAllDescendants(childSpanIds);
-        return allChildSpanIds.includes(each.spanId);
+
+    // Define an array to store the IDs of the span and its descendants (if the span is collapsed)
+    const allRequiredSpanIds = [spanID];
+
+    // If the span is collapsed, recursively find all of its descendants.
+    if (isCollapsed) {
+      const findAllDescendants = (currentChildSpanIds: string[]) => {
+        currentChildSpanIds.forEach(eachId => {
+          const currentChildSpan = trace.spans.find(a => a.spanID === eachId)!;
+          if (currentChildSpan.hasChildren) {
+            allRequiredSpanIds.push(...currentChildSpan.childSpanIds);
+            findAllDescendants(currentChildSpan.childSpanIds);
+          }
+        });
+      };
+      findAllDescendants(allRequiredSpanIds);
+    }
+    const criticalPathSections: criticalPathSection[] = [];
+    // Make a deep copy of the criticalpath to avoid reference errors
+    const criticalPathCopy: criticalPathSection[] = JSON.parse(JSON.stringify(criticalPath));
+    criticalPathCopy.forEach(each => {
+      if (allRequiredSpanIds.includes(each.spanId)) {
```

Comment:
this is confusing - at this point doesn't `criticalPathCopy` contain only segments of the current span? How would segments children spans get there to make this condition true?

Reply:
`criticalPathCopy` means it is just a deep copy of all critical paths.

but anyways I removed it !